### ISSUE TITLE
Skip Style/SymbolProc &:method when refinements are in use

### ISSUE THREAD

**Issue by krissi** [rubocop-hq/rubocop#2720]
When using refinements the conversion `map {|e| e.to_s }` is not the same as `map(&:to_s)`. Rubocop should skip this test when a refinement may be the reason for this code.

Currently the statement would be invalidated by the auto correction as well.

``` ruby
module Refinements
  refine String do
    def to_i
      super * 2
    end
  end
end

class Test
  using Refinements

  def test
    p %w(1 2 3).map(&:to_i)
    p %w(1 2 3).map { |e| e.to_i }
  end
end

Test.new.test
# => [1, 2, 3]
# => [2, 4, 6]
```

Behavior confirmed: https://bugs.ruby-lang.org/issues/9786
Ongoing discussion: https://bugs.ruby-lang.org/issues/9451

**Comment by alexdowad** [comment]
Wow, thanks for pointing this out. Do you think that it is possible for RC to tell whether a refinement could affect `&:method_name` or not? Remember that RC treats each source file separately and does not do global analysis.

**Comment by krissi** [comment]
Hmm, I have no idea how to check this. Without global analysis there is no way I guess. I just verified that the refinements of a class are also applied in the same class in another file (without "using").

Maybe the only way is to disable the autocorrection and disable this check on a per file basis when refinements are used.

**Comment by tim-semba** [comment]
I found a similar issue when a method looks the block's arity.

Example:

```Ruby
def test(&block)
  yield block.arity
end

p test { |e| e.to_s }
p test(&:to_s)

# Result:
#   "1"
#   "-1"
```

Thanks.

**Comment by jonas054** [comment]
> I just verified that the refinements of a class are also applied in the same class in another file (without "using").

@krissi Can you show example code for that? I thought the point of refinements was that they should not be global.

**Comment by krissi** [comment]
@jonas054 Well, it seems I messed something up and your assumption is correct. Sorry for that. So it seems that no global inspection is needed.

Here my testcase which shows that the refinement is file-local only:

refinements.rb:
```ruby
module Refinements
  refine String do
    def to_i
      super * 2
    end
  end
end

```

test.rb:
```ruby
require_relative 'refinements'

class Test
  using Refinements

  def test
    p %w(1 2 3).map(&:to_i)
    p %w(1 2 3).map { |e| e.to_i }
  end
end
```

test2.rb:
```ruby
require_relative 'test'

class Test
  def test2
    p %w(1 2 3).map(&:to_i)
    p %w(1 2 3).map { |e| e.to_i }
  end
end

t = Test.new

t.test
# => [1, 2, 3]
# => [2, 4, 6]

t.test2
# => [1, 2, 3]
# => [1, 2, 3]
```

**Comment by jonas054** [comment]
Then I guess we should take care and at least no auto-correct in this cop if there's a `using` before the method call in question, and if the target ruby version is >= 2.0 (is that when refinements were introduced?).

**Comment by pocke** [comment]
This behaviour has been changed since Ruby 2.4.0.
See https://bugs.ruby-lang.org/issues/9451


```bash
$ ruby -v test2.rb
ruby 2.3.5p376 (2017-09-14 revision 59905) [x86_64-linux]
[1, 2, 3]
[2, 4, 6]
[1, 2, 3]
[1, 2, 3]

$ ruby -v test2.rb
ruby 2.4.0p0 (2016-12-24 revision 57164) [x86_64-linux]
[2, 4, 6]
[2, 4, 6]
[1, 2, 3]
[1, 2, 3]
```

So this cop can auto-correct always with Ruby 2.4 or higher.

**Comment by Darhazer** [comment]
There hasn't been much activity on this ticket and our Core Team is spread too thin on so many tasks, so I'll just close it for the sake of having a cleaner lists of tasks to focus on. We'd gladly review a PR, but it's unlikely that'd we're going to tackle this ourselves in the foreseeable future.
